token/cache.go

```diff
@@ -0,0 +1,93 @@
+package token
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	maxLockWait = 30 * time.Second
+)
+
+// Cache is a thread-safe cache of a authorization token
+// that may be used across http and grpc clients
+type Cache struct {
+	TimeNow func() time.Time
+
+	refreshToken string
+	fetcher      Fetcher
+
+	lock  chan struct{}
+	token atomic.Value
+}
+
+type Fetcher func(ctx context.Context, refreshToken string) (*Token, error)
+
+type Token struct {
+	Bearer  string
+	Expires time.Time
+}
+
+func (t *Token) ExpiresAfter(tm time.Time) bool {
+	return t != nil && t.Expires.After(tm)
+}
+
+func NewCache(fetcher Fetcher, refreshToken string) *Cache {
+	return &Cache{
+		lock:         make(chan struct{}, 1),
```

Comment:
Just to make sure I understand this locking mechanism:
 - we have a channel with a buffer of size 1
 - we "take" the lock if we can successfully send to this channel
 - we "release" the lock by receiving from this channel
 
And therefore only one goroutine can hold the lock at any given time?

Reply:
Correct